File: src/Monads/Either.js
const Right = x => ({
  // map :: Functor f => f a ~> (a -> b) -> f b
  map: f => x |> f |> Right,
  // flatMap :: Functor f => f a ~> (a -> f b) -> f b
  flatMap: f => x |> f,
  // ap :: Functor f => f (a -> b) ~> f a -> f b
  ap: functor => functor.map(x),
  // fold :: Right r => r b ~> (a -> c) -> (b -> c) -> c
  fold: (_f, g) => x |> g,
  // inspect :: Right r => r a ~> () -> String
  inspect: () => `Right(${x})`,
})

const Left = x => ({
  // map :: Functor f => f a ~> (a -> b) -> f a
  map: _f => x |> Left,
  // flatMap :: Functor f => f a ~> (a -> f b) -> f a
  flatMap: _f => x |> Left,
  // ap :: Functor f => f (a -> b) ~> f a -> f (a -> b)
  ap: _functor => x |> Left,
  // fold :: Left l => l a ~> (a -> c) -> (b -> c) -> c
  fold: (f, _g) => x |> f,
  // inspect :: Left l => l a ~> () -> String
  inspect: () => `Left(${x})`,
})

const Either = Right
Either.Right = Right
Either.Left = Left
Either.pure = Either

export default Either

File: src/Monads/Either.hs
import Prelude hiding (Either(..))
import Data.Char

data Either a b = Left a | Right b deriving (Show)

instance Functor (Either e) where
  -- fmap :: Functor f => (a -> b) -> f a -> f b
  fmap f (Right x) = Right (f x)
  fmap _ (Left x) = Left x

instance Applicative (Either e) where
  pure = Right
  Left e <*> _ = Left e
  Right f <*> right = fmap f right

instance Monad (Either e) where
  Left e >>= _ = Left e 
  Right x >>= f = f x



main :: IO ()
main = do
  print (Right 2 :: Either () Int)

